py/vtdb/grpc_vtgate_client.py

```diff
@@ -0,0 +1,536 @@
+        _convert_bind_vars(bind_variables, request.query.bind_variables)
+        it = self.stub.StreamExecuteKeyspaceIds(request, self.timeout)
+
+      elif keyranges is not None:
+        request = vtgate_pb2.StreamExecuteKeyRangesRequest(
+            query=query_pb2.BoundQuery(sql=sql),
+            tablet_type=topodata_pb2.TabletType.Value(tablet_type.upper()),
+            keyspace=keyspace_name,
+        )
+        _add_caller_id(request, effective_caller_id)
+        _add_key_ranges(request, keyranges)
+        _convert_bind_vars(bind_variables, request.query.bind_variables)
+        it = self.stub.StreamExecuteKeyRanges(request, self.timeout)
+
+      else:
+        request = vtgate_pb2.StreamExecuteRequest(
+            query=query_pb2.BoundQuery(sql=sql),
+            tablet_type=topodata_pb2.TabletType.Value(tablet_type.upper()),
+        )
+        _add_caller_id(request, effective_caller_id)
+        _convert_bind_vars(bind_variables, request.query.bind_variables)
+        it = self.stub.StreamExecute(request, self.timeout)
+
+      first_response = it.next()
+    except (face.AbortionError, vtgate_utils.VitessError) as e:
+      self.logger_object.log_private_data(bind_variables)
+      raise _convert_exception(
+          e, sql, keyspace_ids, keyranges,
+          keyspace=keyspace_name, tablet_type=tablet_type)
+
+    fields = []
+    conversions = []
+    for field in first_response.result.fields:
+      fields.append((field.name, field.type))
+      conversions.append(field_types_proto3.conversions.get(field.type))
+
+    def row_generator():
+      try:
+        for response in it:
+          for row in response.result.rows:
+            yield tuple(_make_row(row, conversions))
+      except Exception:
+        logging.exception('gRPC low-level error')
+        raise
+
+    return row_generator(), fields
+
+  def get_srv_keyspace(self, name):
+    try:
+      request = vtgate_pb2.GetSrvKeyspaceRequest(
+          keyspace=name,
+      )
+      response = self.stub.GetSrvKeyspace(request, self.timeout)
+      return keyspace.Keyspace(
+          name,
+          keyrange_constants.srv_keyspace_proto3_to_old(response.srv_keyspace))
+
+    except (face.AbortionError, vtgate_utils.VitessError) as e:
+      raise _convert_exception(e, keyspace=name)
+
+  def _add_session(self, request):
+    if self.session:
+      request.session.CopyFrom(self.session)
+
+
+def _add_caller_id(request, caller_id):
+  if caller_id:
+    if caller_id.principal:
+      request.caller_id.principal = caller_id.principal
+    if caller_id.component:
+      request.caller_id.component = caller_id.component
+    if caller_id.subcomponent:
+      request.caller_id.subcomponent = caller_id.subcomponent
+
+
+def _add_key_ranges(request, keyranges):
+  for kr in keyranges:
+    encoded_kr = request.key_ranges.add()
+    encoded_kr.start = kr.Start
+    encoded_kr.end = kr.End
+
+
+def _extract_rpc_error(exec_method, error):
+  if error.code:
```

Comment:
Wait, are you sure this is doing the right thing?

_extract_rpc_error should be returning errors for partial errors - i.e., for errors that are embedded in the regular response returned by VtGate RPC calls. I assume error.code only exists for when errors are returned without a response.